Introduce an optional rdfCollectionRepresentation attribute on ResourceProperties which can be set to rdfList or rdfBag (or n/a)

jadelkhoury opened the pull request:
Introduce an optional rdfCollectionRepresentation attribute on ResourceProperties which can be set to rdfList or rdfBag (or n/a)

**NOTE:** This is what I have tested. But I have NOT tested that the actual annotation produces the expected result from JenaModelHelper.

* if set to "n_a", no affect on produced java code.
* if set on properties with cardinality 0..1 or 1: no effect on produced
hava code.

* If set to "RDFBag" on a Property with a "many" cardinality:
	1. We add the @OslcRdfCollectionType(collectionType =
OslcRdfCollectionType.RDF_BAG).
	2. The property is defined as Set/Hashset in the Java code.

* if set to "RDFList" on a Property with a "many" cardinality:

	1. We add the @OslcRdfCollectionType(collectionType =
OslcRdfCollectionType.RDF_LIST).
	2. The property is defined as List/ArrayList in the Java code.

jadelkhoury commented:
@berezovskyi Can you safely review this without me sharing Java generated code?

Ignore the commited java files (from EMF). Main logic is in the resourcePropertyServices.mtl file.

berezovskyi commented on org.eclipse.lyo.oslc4j.codegenerator/src/org/eclipse/lyo/oslc4j/codegenerator/services/resourcePropertyServices.mtl:
```diff
@@ -69,178 +69,215 @@ endif)
+[query private javaAttributeCollectionType(aProperty: ResourceProperty) : String =
+(if ((not aProperty.rdfCollectionRepresentation.oclIsUndefined()) and (not aProperty.occurs.oclIsUndefined())) then
+    if (aProperty.occurs = ResourcePropertyOccurs::zeroOrMany or aProperty.occurs = ResourcePropertyOccurs::oneOrMany) then
+        if (aProperty.rdfCollectionRepresentation = ResourcePropertyRdfCollectionRepresentation::rdfList) then
```
or rdfSeq

berezovskyi commented on org.eclipse.lyo.tools.adaptormodel.edit/plugin.properties:
```diff
@@ -220,3 +220,7 @@ _UI_MavenProjectConfiguration_groupId_feature = Group Id
 _UI_MavenServerConfiguration_jettyPort_feature = Jetty Port
+_UI_ResourceProperty_rdfCollectionRepresentation_feature = Rdf Collection Representation
+_UI_ResourcePropertyRdfCollectionRepresentation_n_a_literal = n_a
+_UI_ResourcePropertyRdfCollectionRepresentation_rdfList_literal = rdfList
```
how about 'rdf:List' or 'RDF List'

berezovskyi commented on org.eclipse.lyo.tools.adaptormodel.edit/plugin.properties:
```diff
@@ -220,3 +220,7 @@ _UI_MavenProjectConfiguration_groupId_feature = Group Id
 _UI_MavenProjectConfiguration_version_feature = Version
 _UI_MavenServerConfiguration_jettyPort_feature = Jetty Port
+_UI_ResourceProperty_rdfCollectionRepresentation_feature = Rdf Collection Representation
+_UI_ResourcePropertyRdfCollectionRepresentation_n_a_literal = n_a
```
how about 'Default' or 'Default (repeated predicate)' or 'Default (repeated property)'

jadelkhoury replied:
you mean instead of "n_a", maybe "none" makes more sense. It is not about default. It's about saying there is no RDF Collection.

berezovskyi replied:
No, here I mean how do I want to serialise a property with ..n cardinality:

- the Default way (just repeat the property n times for a Set with n items)
- use RDF Collection (List)
- use RDF Seq Container
- use RDF Bag Container

jadelkhoury replied:
I see your point.
Let me think about it.

berezovskyi commented on org.eclipse.lyo.tools.adaptormodel.edit/plugin.properties:
```diff
@@ -220,3 +220,7 @@ _UI_MavenProjectConfiguration_groupId_feature = Group Id
+_UI_ResourceProperty_rdfCollectionRepresentation_feature = Rdf Collection Representation
+_UI_ResourcePropertyRdfCollectionRepresentation_n_a_literal = n_a
+_UI_ResourcePropertyRdfCollectionRepresentation_rdfList_literal = rdfList
+_UI_ResourcePropertyRdfCollectionRepresentation_rdfBag_literal = rdfBag
```
how about 'rdf:Bag' or "RDF 'Bag' Container"

jadelkhoury replied:
ok. I'll fix the display.

berezovskyi commented on org.eclipse.lyo.oslc4j.codegenerator/src/org/eclipse/lyo/oslc4j/codegenerator/services/resourcePropertyServices.mtl:
```diff
@@ -69,178 +69,215 @@ endif)
+            '@OslcRdfCollectionType(collectionType = OslcRdfCollectionType.RDF_LIST)'
+        else
+            if (aProperty.rdfCollectionRepresentation = ResourcePropertyRdfCollectionRepresentation::rdfBag) then
+                '@OslcRdfCollectionType(collectionType = OslcRdfCollectionType.RDF_BAG)'
```
stictly speaking, RDF List is an RDF Collection and RDF Bag or Seq are RDF _Containers_ but its OK.

jadelkhoury commented on org.eclipse.lyo.oslc4j.adaptormodel/model/adaptorInterface.ecore:
```diff
@@ -177,6 +179,11 @@
     <eLiterals name="false" literal="false"/>
     <eLiterals name="true" value="1" literal="true"/>
   </eClassifiers>
+  <eClassifiers xsi:type="ecore:EEnum" name="ResourcePropertyRdfCollectionRepresentation">
```
Unless we can test and need to use Bag, I'd rather remove it then.

jadelkhoury commented:
@nilsjor @berezovskyi 
Please review 2nd commit. 
I removed support for RDF:Bag. I also renamed "n_a" to "none".
It was not easy to re-label the options to something more intuitive. I would rather do this for many other parameters at once.

jadelkhoury commented:
@berezovskyi 
What do you think of this naming? Last offer.

![image](https://user-images.githubusercontent.com/7108694/78299773-bbf06880-7536-11ea-9cad-2bf7811e6ad2.png)

berezovskyi commented:
Sounds good!
